aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,274 @@
+        .expect("Peer id needs to be given");
+
+    // Define set of schema ids we are interested in
+    let supported_schema_ids: Vec<SchemaId> = context
+        .schema_provider
+        .all()
+        .await
+        .iter()
+        .map(|schema| schema.id().to_owned())
+        .collect();
+    let target_set = TargetSet::new(&supported_schema_ids);
+
+    // Run a connection manager which deals with the replication logic
+    let manager = ConnectionManager::new(
+        &context.schema_provider,
+        &context.store,
+        &tx,
+        local_peer_id,
+        target_set,
+    );
+
+    let handle = task::spawn(manager.run());
+
+    if tx_ready.send(()).is_err() {
+        warn!("No subscriber informed about replication service being ready");
+    };
+
+    tokio::select! {
+        _ = handle => (),
+        _ = shutdown => {
+            // @TODO: Wait until all pending replication processes are completed during graceful
+            // shutdown
+        }
+    }
+
+    Ok(())
+}
+
+struct PeerStatus {
+    peer_id: PeerId,
+    successful_count: usize,
+    failed_count: usize,
+}
+
+impl PeerStatus {
+    pub fn new(peer_id: &PeerId) -> Self {
+        Self {
+            peer_id: *peer_id,
+            successful_count: 0,
+            failed_count: 0,
+        }
+    }
+}
+
+struct ConnectionManager {
+    peers: HashMap<PeerId, PeerStatus>,
+    sync_manager: SyncManager<PeerId>,
+    tx: ServiceSender,
+    rx: Receiver<ServiceMessage>,
+    target_set: TargetSet,
+}
+
+impl ConnectionManager {
+    pub fn new(
+        schema_provider: &SchemaProvider,
+        store: &SqlStore,
+        tx: &ServiceSender,
+        local_peer_id: PeerId,
+        target_set: TargetSet,
+    ) -> Self {
+        let ingest = SyncIngest::new(schema_provider.clone(), tx.clone());
+        let sync_manager = SyncManager::new(store.clone(), ingest, local_peer_id);
+
+        Self {
+            peers: HashMap::new(),
+            sync_manager,
+            tx: tx.clone(),
+            rx: tx.subscribe(),
+            target_set,
+        }
+    }
+
+    async fn on_connection_established(&mut self, peer_id: PeerId) {
+        if self
+            .peers
+            .insert(peer_id, PeerStatus::new(&peer_id))
+            .is_some()
+        {
+            warn!("Duplicate established connection encountered");
+        }
+
+        self.update_sessions().await;
+    }
+
+    async fn on_connection_closed(&mut self, peer_id: PeerId) {
+        // Clear running replication sessions from sync manager
+        self.sync_manager.remove_sessions(&peer_id);
+
+        // Remove peer from our connections table
+        if self.peers.remove(&peer_id).is_none() {
+            warn!("Tried to remove unknown connection");
+        }
+
+        self.update_sessions().await;
+    }
+
+    async fn on_replication_message(&mut self, peer_id: PeerId, message: SyncMessage) {
+        match self.sync_manager.handle_message(&peer_id, &message).await {
```

Comment:
Probably want to make the `IngestError` transparent so we understand better why arriving entries do not get accepted:

```
#[error(transparent)]
```

Reply:
I was getting a `ValidationError::SeqNumDoesNotMatch(1, 2)` error when logging this, but it's gone away now since i made some changes. Not sure _exactly_ what was causing it, but making sure to send the system schema over first definitely helped.